### ISSUE TITLE
Flooded with warnings

### ISSUE THREAD

**Issue by renambot-uic** [tmarrinan/websocketio#2]
Hi ;-)
In SAGE2, now I get flooded with warnings 'recipient has no listener...' when we had only the occasional warning before.
I see you have a recent commit about a 1sec delay for retrying.
I think we flood the socket with messages faster than they can be deleted.
Maybe an intermediate solution would be a parameter 'retry' with a default value of 'false', to maintain the old behavior...

Thanks

**Comment by tmarrinan** [comment]
It should be working properly, but now may be printing that on the
broadcast command in addition to the direct emit. I can modify that or add
a 'quiet' mode. Are there any errors happening in SAGE2?
Thanks for letting me know.

Tommy Marrinan

**Comment by renambot-uic** [comment]
The problem is if you send a given message type at 20fps without receiver, the queue never empties.
So your update removes the print messages, but it keeps trying.

Also, line 218, it should be 'this' not '_this'

**Comment by tmarrinan** [comment]
Thanks for catching the typo.
True the queue wouldn't ever be empty, but it wasn't before either. It
shouldn't be constantly growing either though. Data should be added and
removed at the same rate. Are you noticing a performance decrease?

Tommy Marrinan

**Comment by renambot-uic** [comment]
I think it was different before because, in broadcast, you were doing a 'remoteListeners' check upfront, which would remove any check later in the send. So it was lighter weight I think, and not keeping a copy of the messages.

**Comment by tmarrinan** [comment]
Yes, I did change that. Previously, I tried every 16 milliseconds to resend
a message if there was no listener (give up after 16 tries). Now I save a
copy of the message if there is no listener, then send it once that
listener is registered or delete the message after waiting 1 second. I have
to do one of the two because a websocket message may attempt to send before
receiving a message that registers the listener.
I did do a remoteListeners check upfront in the broadcast - however this
actually led to errors. If I did a broadcast right when a new client
connected, then the new client may miss receiving that broadcast message.

The only impact I would anticipate this having is a higher memory footprint
when constantly broadcasting a message. Are you noticing any performance
hits?

Tommy Marrinan
